### PR TITLE
Add Apollo test of restore a replica from a snapshot of other replica

### PR DESCRIPTION
**Pull request by Efrat1**
Also format files.

### CODE REVIEW

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 1]
```diff
@@ -846,7 +846,52 @@ async def test_signed_state_snapshot_read_as_of_req_non_existent_checkpoint(self
...
+
+        crashed_replica = list(bft_network.random_set_of_replicas(1, {initial_prim}))[0]
+        bft_network.stop_replica(crashed_replica)
+        for i in range(150):  # run till the next checkpoint
+            await skvbc.send_write_kv_set()
+
+        src_replica = list(bft_network.random_set_of_replicas(1, without={crashed_replica}))
+        await self.wait_for_stable_checkpoint(bft_network, src_replica, 300)
+        await self.wait_for_created_snapshots_metric(bft_network, src_replica, 2)
+        snapshot_id = await bft_network.get_metric(src_replica[0], bft_network,
```
Comment (WildFireFlum) [inline tests/apollo/test_skvbc_dbsnapshot.py:873]: This metric is used by many tests, you can extract a function for it

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 2]
```diff
@@ -846,7 +846,52 @@ async def test_signed_state_snapshot_read_as_of_req_non_existent_checkpoint(self
...
+        self.reset_metadata(bft_network, crashed_replica)
+        bft_network.start_replica(crashed_replica)
+
+        # make sure that the restored replica participates in consensus
+        for i in range(150):
+            await skvbc.send_write_kv_set()
+
+        non_crash_reps = bft_network.all_replicas(without={crashed_replica})
+        for r in non_crash_reps:
+            nb_fast_path = await bft_network.get_metric(r, bft_network, "Counters", "totalFastPaths")
```
Comment (WildFireFlum) [inline tests/apollo/test_skvbc_dbsnapshot.py:889]: use num_of_fast_path_requests

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 3]
```diff
@@ -846,7 +846,52 @@ async def test_signed_state_snapshot_read_as_of_req_non_existent_checkpoint(self
...
+            self.assertGreater(nb_fast_path, fast_paths[r])
+            fast_paths[r] = (fast_paths[r], nb_fast_path)
+        nb_fast_path = await bft_network.get_metric(crashed_replica, bft_network, "Counters", "totalFastPaths")
```
Comment (WildFireFlum) [inline tests/apollo/test_skvbc_dbsnapshot.py:892]: use num_of_fast_path_requests

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 4]
```diff
@@ -956,3 +1001,24 @@ async def validate_stop_on_wedge_point(self, bft_network, skvbc, fullWedge=False
                     if stopped_replicas < stop_condition:
                         done = False
 
+    def reset_metadata(self, bft_network, replica_id):
```
Comment (WildFireFlum) [inline tests/apollo/test_skvbc_dbsnapshot.py:1004]: log this operation

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 1 above)
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:873]: done

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 2 above)
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:889]: done

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 3 above)
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:892]: done

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 4 above)
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:1004]: done, thanks

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 5]
```diff
@@ -846,7 +833,51 @@ async def test_signed_state_snapshot_read_as_of_req_non_existent_checkpoint(self
...
+
+    @with_trio
+    @with_bft_network(start_replica_cmd, selected_configs=lambda n, f, c: n == 7)
+    @verify_linearizability()
+    async def test_restore_from_snapshot_of_other(self, bft_network, tracker):
+        initial_prim = 0
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network, tracker)
+        for i in range(DB_CHECKPOINT_WIN_SIZE):  # get to a checkpoint
+            await skvbc.send_write_kv_set()
```
Comment (WildFireFlum) [inline tests/apollo/test_skvbc_dbsnapshot.py:845]: There is a function in skvbc_commit_path called send_kvs_sequentially which does the same and logs the index of every sent kv.
You can move it to the ApolloTest class and call it using self.send_kvs_sequentially.

Another thing (which might be ever clearer) is to add a function of your own send_until_checkpoints_reached(checkpoint_count) which calls send_kvs_sequentially.

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 6]
```diff
@@ -346,8 +345,7 @@ async def test_create_dbcheckpoint_with_parallel_client_requests(self, bft_netwo
             self.assertTrue(any(dbcheckpoint_info.seq_num ==
                             300 for dbcheckpoint_info in dbcheckpoint_info_list))
         for replica_id in bft_network.all_replicas():
-            last_blockId = await bft_network.get_metric(replica_id, bft_network,
-                                                       "Gauges", "lastDbCheckpointBlockId", component="rocksdbCheckpoint")
+            last_blockId = await bft_network.get_metric(replica_id)
```
[the remark this answers is not in the document]
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:348]: right, thanks

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 7]
```diff
@@ -698,8 +688,8 @@ async def test_signed_public_state_hash_req_existing_checkpoint(self, bft_networ
         for ser_rsi in ser_rsis.values():
             rsi_resp = cmf_msgs.ReconfigurationResponse.deserialize(ser_rsi)[0]
             self.assertEqual(rsi_resp.response.status, cmf_msgs.SnapshotResponseStatus.Success)
-            self.assertEqual(rsi_resp.response.data.snapshot_id, 600)
-            self.assertEqual(rsi_resp.response.data.block_id, 600)
+            self.assertEqual(rsi_resp.response.data.snapshot_id, DB_CHECKPOINT_HIGH_WIN_SIZE)
+            self.assertEqual(rsi_resp.response.data.block_id, DB_CHECKPOINT_HIGH_WIN_SIZE)
             # Expect the SHA3-256 hash of the empty string.
             self.assertEqual(bytearray(rsi_resp.response.data.hash), bytes.fromhex("a7ffc6f8bf1ed76651c14756a061d662f580ff4de43b49fa82d80a4b80f8434a"))
```
[the remark this answers is not in the document]
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:694]: done

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 8]
```diff
@@ -611,8 +604,7 @@ async def test_db_checkpoint_creation_with_wedge(self, bft_network):
         await self.wait_for_stable_checkpoint(bft_network, bft_network.all_replicas(), 900)
```
[the remark this answers is not in the document]
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:604]: window size in this test is DB_CHECKPOINT_HIGH_WIN_SIZE (600).
I leave the 900 as is.

File: tests/apollo/test_skvbc_dbsnapshot.py (hunk 5 above)
Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py:845]: thanks.
Used send_kvs_sequentially only, as sometimes we don't send kvs until a checkpoint

File: tests/apollo/test_skvbc_dbsnapshot.py [hunk 9]
```diff
@@ -158,27 +161,35 @@ def start_replica_cmd_with_operator_and_public_keys(builddir, replica_id):
...
-            "-j", "600",
+            "-j", str(DB_CHECKPOINT_HIGH_WIN_SIZE),
             "-o", builddir + "/operator_pub.pem",
             "--add-all-keys-as-public"]
 
 class SkvbcDbSnapshotTest(ApolloTest):
     __test__ = False  # so that PyTest ignores this test scenario
 
+    @staticmethod
+    async def send_kvs_sequentially(skvbc: 'SimpleKVBCProtocol', kv_count: int, client: 'BftClient' = None):
```
Comment (WildFireFlum) [inline tests/apollo/test_skvbc_dbsnapshot.py:172]: This is a copy of a method used in another test class, I meant you should move it to the parent, not copy it.

Comment (Efrat1) [reply tests/apollo/test_skvbc_dbsnapshot.py]: Moved to skvbc.py

**Review by toly-kournik — APPROVED** [review]
LGTM!
thanks, Efrat!

**Pull request merged**